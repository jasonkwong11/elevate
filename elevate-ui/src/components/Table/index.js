import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SORTS from '../../constants';
import Sort from '../Sort'

const AlertDetails = ({details}) =>
  <div className="alert-details">
    <h4>Alert Details #{details[0]}</h4>
    <p>Alert Id: {details[0]}</p>
    <p>Alert Time: {details[1]}</p>
    <p>Severity: {details[2]}</p>
    <p>Protocol: {details[3]}</p>
    <p>Client IP: {details[4]}</p>
    <p>Server IP: {details[5]}</p>
    <p>Client Country: {details[6]}</p>
  </div>

class Table extends Component {
  
  constructor(props) {
    super(props);
    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
      showAlertDetails: false,
      clickedAlertDetails: []
    };
    this.onSort = this.onSort.bind(this);
    this.onButtonClick = this.onButtonClick.bind(this);
  }

  onSort(sortKey) {
    this.setState((prevState) => {
      const isSortReverse = prevState.sortKey === sortKey && !prevState.isSortReverse;
      return {
        sortKey,
        isSortReverse 
      }
    });
  }

  onButtonClick(e) {
    let details = e.currentTarget.innerText.split("\n");

    this.setState({
      showAlertDetails: !this.state.showAlertDetails,
      clickedAlertDetails: details
    });
  }

  render() {
    const {
      alerts
    } = this.props;

    const {
      sortKey,
      isSortReverse,
      clickedAlertDetails
    } = this.state;

    const sortedList = SORTS[sortKey](alerts);
    const reverseSortList = isSortReverse
      ? sortedList.reverse()
      : sortedList

    return(
      <div className="table">
        {this.state.showAlertDetails ?
           <AlertDetails 
            details={clickedAlertDetails}
           /> :
           null
        }
          <div className="table-header">
            <span style={smallColumn}>
              <Sort
                sortKey={'ALERTID'}
                onSort={this.onSort}
                activeSortKey={sortKey}
              >
                AlertId
              </Sort>
            </span>
            <span style={midColumn}>
              <Sort
                sortKey={'ALERTTIME'}
                onSort={this.onSort}
                activeSortKey={sortKey}
              >
                AlertTime
              </Sort>
            </span>
            <span style={smallColumn}>
              <Sort
                sortKey={'SEVERITY'}
                onSort={this.onSort}
                activeSortKey={sortKey}
              >
                Severity
              </Sort>
            </span>
            <span style={midColumn}>
              <Sort
                sortKey={'CLIENTIP'}
                onSort={this.onSort}
                activeSortKey={sortKey}
              >
                ClientIP
              </Sort>
            </span>
            <span style={midColumn}>
              <Sort
                sortKey={'SERVERIP'}
                onSort={this.onSort}
                activeSortKey={sortKey}
              >
                ServerIP
              </Sort>
            </span>
            <span style={smallColumn}>
              <Sort
                sortKey={'PROTOCOL'}
                onSort={this.onSort}
                activeSortKey={sortKey}
              >
                Protocol
              </Sort>
            </span>
            <span style={midColumn}>
              <Sort
                sortKey={'CLIENTCOUNTRY'}
                onSort={this.onSort}
                activeSortKey={sortKey}
              >
                ClientCountry
              </Sort>
            </span>
          </div>
          { reverseSortList.map(item =>
            <div key={item["AlertId"]} className="table-row" onClick={this.onButtonClick}>
              <span style={smallColumn}>
                {item.AlertId}
              </span>
              <span style={midColumn}>
                {item.AlertTime}
              </span>
              <span style={smallColumn}>
                {item.Severity}
              </span>
              <span style={midColumn}>
                {item.ClientIP}
              </span>
              <span style={midColumn}>
                {item.ServerIP}
              </span>
              <span style={smallColumn}>
                {item.Protocol}
              </span>
              <span style={midColumn}>
                {item.ClientCountry}
              </span>
            </div>
          )}
        </div>
    );
  }
}

Table.propTypes = {
  alerts: PropTypes.arrayOf(
    PropTypes.shape({
      alertId: PropTypes.number,
      alertTime: PropTypes.string,
      severity: PropTypes.string,
      clientIp: PropTypes.string,
      serverIp: PropTypes.string,
      protocol: PropTypes.string,
      clientCountry: PropTypes.string
    })
  ).isRequired
}

// table column styles
  const midColumn = {
    width: '30%'
  }

  const smallColumn = {
    width: '10%'
  }
// end styles

export default Table