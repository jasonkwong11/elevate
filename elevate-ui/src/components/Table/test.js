import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Table from './index';


describe('Table', () => {

  const props = {
    alerts: [
      {
        "AlertId": 24859,
        "AlertTime": "2018-02-26 07:59:22",
        "Severity": "Low",
        "ClientIP": "155.175.121.134",
        "ServerIP": "154.224.236.156",
        "Protocol": "SMTP",
        "ClientCountry": "Canada"
      },
      {
        "AlertId": 24860,
        "AlertTime": "2018-02-21 15:28:26",
        "Severity": "Medium",
        "ClientIP": "66.78.200.140",
        "ServerIP": "142.6.126.64",
        "Protocol": "HTTP",
        "ClientCountry": "Germany"
      }
    ]
  };

  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  it('shows two items in list', () => {
    const element = shallow(
      <Table { ...props } />
    );

    expect(element.find('.table-row').length).toBe(2);
  })

  test('snapshots', () => {
    const component = renderer.create(
      <Table { ...props } />
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});