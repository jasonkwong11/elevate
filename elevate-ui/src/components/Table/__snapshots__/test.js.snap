// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        AlertId
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        AlertTime
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Severity
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        ClientIP
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        ServerIP
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Protocol
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        ClientCountry
      </button>
    </span>
  </div>
  <div
    className="table-row"
    onClick={[Function]}
  >
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      24859
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2018-02-26 07:59:22
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Low
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      155.175.121.134
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      154.224.236.156
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      SMTP
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Canada
    </span>
  </div>
  <div
    className="table-row"
    onClick={[Function]}
  >
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      24860
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2018-02-21 15:28:26
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Medium
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      66.78.200.140
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      142.6.126.64
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      HTTP
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Germany
    </span>
  </div>
</div>
`;
